flashlight/fl/tensor/backend/jit/JitTensorBase.cpp

```diff
@@ -15,44 +15,56 @@
 
 namespace fl {
 
-JitTensorBase::JitTensorBase(Node* node) : node_(node) {
-  node_->incRefCount();
-}
+struct JitTensorBase::ShareData {
+  Node* node;
+  ShareData(Node* node) : node(node) {
+    node->incRefCount();
+  }
+  ~ShareData() {
+    node->decRefCount();
+  }
+};
 
-JitTensorBase::~JitTensorBase() {
-  node_->decRefCount();
-}
+JitTensorBase::JitTensorBase(Node* node)
+    : JitTensorBase(std::make_shared<ShareData>(node)) {}
+
+JitTensorBase::JitTensorBase(std::shared_ptr<ShareData> sharedNode)
+    : sharedNode_(sharedNode) {}
+
+JitTensorBase::~JitTensorBase() {}
 
-// take care of refcount for old & new nodes
 void JitTensorBase::replaceNode(Node* newNode) const {
-  if (node_ != newNode) {
-    node_->decRefCount();
+  auto oldNode = node();
+  if (newNode != oldNode) {
     newNode->incRefCount();
-    const_cast<JitTensorBase*>(this)->node_ = newNode;
+    oldNode->decRefCount();
+    sharedNode_->node = newNode;
```

Comment:
if you did use shared_ptr, some built-in compiler optimizations can actually avoid a refcount change here
```cpp
replaceNode(std::shared_ptr<Node>& newNode) const {
  node_.swap(newNode);
}
```